File: backend/src/aerospike_cluster_manager_api/models/connection.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class ConnectionStatus(BaseModel):
    connected: bool
    nodeCount: int
    namespaceCount: int
    build: str | None = None
    edition: str | None = None
    totalOps: int = 0
    memoryUsed: int = 0
    memoryTotal: int = 0
    diskUsed: int = 0
    diskTotal: int = 0


class ConnectionProfile(BaseModel):
    id: str
    name: str
    hosts: list[str] = Field(min_length=1)
    port: int = Field(ge=1, le=65535)
    clusterName: str | None = None
    username: str | None = None
    password: str | None = None
    color: str = Field(pattern=r"^#[0-9a-fA-F]{6}$")
    label: str | None = None
    label_color: str | None = None
    description: str | None = None
    createdAt: str
    updatedAt: str


class CreateConnectionRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(min_length=1, max_length=255, default="New Connection")
    hosts: list[str] = Field(min_length=1, default=["localhost"])
    port: int = Field(ge=1, le=65535, default=3000)
    clusterName: str | None = None
    username: str | None = None
    password: str | None = None
    color: str = Field(pattern=r"^#[0-9a-fA-F]{6}$", default="#0097D3")
    label: str | None = None
    label_color: str | None = Field(None, alias="labelColor", pattern=r"^#[0-9a-fA-F]{6}$")
    description: str | None = None


class UpdateConnectionRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str | None = Field(None, min_length=1, max_length=255)
    hosts: list[str] | None = Field(None, min_length=1)
    port: int | None = Field(None, ge=1, le=65535)
    clusterName: str | None = None
    username: str | None = None
    password: str | None = None
    color: str | None = Field(None, pattern=r"^#[0-9a-fA-F]{6}$")
    label: str | None = None
    label_color: str | None = Field(None, alias="labelColor", pattern=r"^#[0-9a-fA-F]{6}$")
    description: str | None = None


class TestConnectionRequest(BaseModel):
    hosts: list[str] = Field(min_length=1)
    port: int = Field(ge=1, le=65535, default=3000)
    username: str | None = None
    password: str | None = None


class ConnectionProfileResponse(BaseModel):
    """Connection profile without password — used in API responses."""

    id: str
    name: str
    hosts: list[str] = Field(min_length=1)
    port: int = Field(ge=1, le=65535)
    clusterName: str | None = None
    username: str | None = None
    color: str = Field(pattern=r"^#[0-9a-fA-F]{6}$")
    label: str | None = None
    labelColor: str | None = None
    description: str | None = None
    createdAt: str
    updatedAt: str

    @classmethod
    def from_profile(cls, profile: ConnectionProfile) -> ConnectionProfileResponse:
        data = profile.model_dump(exclude={"password", "label_color"})
        data["labelColor"] = profile.label_color
        return cls(**data)


class ConnectionWithStatus(ConnectionProfileResponse):
    status: ConnectionStatus

File: ui/src/lib/types/connection.ts
/**
 * Connection-related types mirrored from API Pydantic models.
 * See: api/src/aerospike_cluster_manager_api/models/connection.py
 */

/**
 * Why a health check reported `connected: false`.
 *
 * Deliberately coarse (#470). The health route is reachable unauthenticated in
 * the default configuration, so the older set — `timeout`,
 * `connection_refused`, `cluster_error`, `unknown` — let a caller distinguish a
 * closed port from a filtered one from a live non-Aerospike listener, i.e. it
 * was a port scanner. Those four collapse into `unreachable`.
 */
export type ConnectionErrorType =
  | "not_found"
  | "unreachable"
  | "auth_error"
  | "blocked_target"

export interface ConnectionStatus {
  connected: boolean
  nodeCount: number
  namespaceCount: number
  build?: string | null
  edition?: string | null
  memoryUsed: number
  memoryTotal: number
  diskUsed: number
  diskTotal: number
  tendHealthy?: boolean | null
  error?: string | null
  errorType?: ConnectionErrorType | null
}

export interface ConnectionProfileResponse {
  id: string
  name: string
  hosts: string[]
  port: number
  clusterName?: string | null
  username?: string | null
  color: string
  note?: string | null
  /** Always contains an ``env`` key after backend normalization (defaults to ``default``). */
  labels: Record<string, string>
  /** Workspace this connection belongs to. Defaults to ``ws-default`` server-side. */
  workspaceId: string
  createdAt: string
  updatedAt: string
}

export interface ConnectionWithStatus extends ConnectionProfileResponse {
  status: ConnectionStatus
}

export interface CreateConnectionRequest {
  name?: string
  hosts?: string[]
  port?: number
  clusterName?: string | null
  username?: string | null
  password?: string | null
  color?: string
  note?: string | null
  labels?: Record<string, string> | null
  /** When omitted, the backend assigns the connection to the default workspace. */
  workspaceId?: string | null
}

export interface UpdateConnectionRequest {
  name?: string
  hosts?: string[]
  port?: number
  clusterName?: string | null
  username?: string | null
  password?: string | null
  color?: string
  note?: string | null
  labels?: Record<string, string> | null
  /** Set to move the connection into a different workspace. */
  workspaceId?: string | null
}

export interface TestConnectionRequest {
  hosts: string[]
  port?: number
  username?: string | null
  password?: string | null
}

export interface TestConnectionResponse {
  success: boolean
  message: string
}
